provider/aws_sd.go

```diff
@@ -0,0 +1,665 @@
+	if err != nil {
+		return nil, err
+	}
+
+	for _, ns := range namespaces {
+		services, err := p.ListServicesByNamespaceID(ns.Id)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, srv := range services {
+			instances, err := p.ListInstancesByServiceID(srv.Id)
+			if err != nil {
+				return nil, err
+			}
+
+			if len(instances) > 0 {
+				// DNS name of the record is a concatenation of service and namespace
+				dnsName := *srv.Name + "." + *ns.Name
+				ep := p.instancesToEndpoint(dnsName, srv, instances)
+				endpoints = append(endpoints, ep)
+			}
+		}
+	}
+
+	return endpoints, nil
+}
+
+func (p *AWSSDProvider) instancesToEndpoint(recordName string, srv *sd.Service, instances []*sd.InstanceSummary) *endpoint.Endpoint {
+	newEndpoint := &endpoint.Endpoint{
+		DNSName:   recordName,
+		RecordTTL: endpoint.TTL(aws.Int64Value(srv.DnsConfig.DnsRecords[0].TTL)),
+		Targets:   make(endpoint.Targets, 0, len(instances)),
+	}
+
+	for _, inst := range instances {
+		// CNAME
+		if inst.Attributes[sdInstanceAttrCname] != nil && aws.StringValue(srv.DnsConfig.DnsRecords[0].Type) == sd.RecordTypeCname {
+			newEndpoint.RecordType = endpoint.RecordTypeCNAME
+			newEndpoint.Targets = append(newEndpoint.Targets, aws.StringValue(inst.Attributes[sdInstanceAttrCname]))
+
+			// ALIAS
+		} else if inst.Attributes[sdInstanceAttrAlias] != nil {
+			newEndpoint.RecordType = endpoint.RecordTypeCNAME
+			newEndpoint.Targets = append(newEndpoint.Targets, aws.StringValue(inst.Attributes[sdInstanceAttrAlias]))
+
+			// IP-based target
+		} else if inst.Attributes[sdInstanceAttrIPV4] != nil {
+			newEndpoint.RecordType = endpoint.RecordTypeA
+			newEndpoint.Targets = append(newEndpoint.Targets, aws.StringValue(inst.Attributes[sdInstanceAttrIPV4]))
+		} else {
+			log.Warnf("Invalid instance \"%v\" found in service \"%v\"", inst, srv.Name)
+		}
+	}
+
+	return newEndpoint
+}
+
+// ApplyChanges applies Kubernetes changes in endpoints to AWS API
+func (p *AWSSDProvider) ApplyChanges(changes *plan.Changes) error {
+	// return early if there is nothing to change
+	if len(changes.Create) == 0 && len(changes.Delete) == 0 && len(changes.UpdateNew) == 0 {
+		log.Info("All records are already up to date")
+		return nil
+	}
+
+	// convert updates to delete and create operation if applicable (updates not supported)
+	creates, deletes := p.updatesToCreates(changes)
+	changes.Delete = append(changes.Delete, deletes...)
+	changes.Create = append(changes.Create, creates...)
+
+	namespaces, err := p.ListNamespaces()
+	if err != nil {
+		return err
+	}
+
+	// Deletes must be executed first to support update case.
+	// When just list of targets is updated `[1.2.3.4] -> [1.2.3.4, 1.2.3.5]` it is translated to:
+	// ```
+	// deletes = [1.2.3.4]
+	// creates = [1.2.3.4, 1.2.3.5]
+	// ```
+	// then when deletes are executed after creates it will miss the `1.2.3.4` instance.
+	err = p.submitDeletes(namespaces, changes.Delete)
+	if err != nil {
+		return err
+	}
+
```

Comment:
i think we need to let user know that in certain cases, e.g. (ungracefuly termination of external-dns), might lead to a temporary deletion of all records